opendevin/controller/__init__.py

```diff
@@ -0,0 +1,77 @@
+from typing import List, Callable
+
+from opendevin.lib.event import Event
+from opendevin.state import State
+from opendevin.agent import Agent
+from opendevin.action import Action
+
+from .command_manager import CommandManager
+
+
+def print_callback(event):
+    print(event, flush=True)
+
+
+class AgentController:
+    def __init__(
+        self,
+        agent: Agent,
+        workdir: str,
+        max_iterations: int = 100,
+        callbacks: List[Callable] = [],
+    ):
+        self.agent = agent
+        self.max_iterations = max_iterations
+        self.background_commands = []
+        self.command_manager = CommandManager(workdir)
+
+        self.callbacks = callbacks
+        self.callbacks.append(self.agent.add_event)
+        self.callbacks.append(print_callback)
+
+    def maybe_perform_action(self, event):
+        if not (event and event.is_runnable()):
+            return
+        action = "output"
+        try:
+            output = event.run(self)
+        except Exception as e:
+            output = "Error: " + str(e)
+            action = "error"
+        out_event = Event(action, {"output": output})
+        return out_event
+
+    @property
+    def state(self) -> State:
+        return State(
+            background_commands=self.command_manager.background_commands,
+            background_events=self.command_manager.get_background_events(),
+        )
+
+    def start_loop(self):
+        for i in range(self.max_iterations):
+            print("STEP", i, flush=True)
+
+            # NOTE: Backgorund events are now part of the State
+            # log_events = self.command_manager.get_background_events()
+            # for event in log_events:
+            #     for callback in self.callbacks:
+            #         callback(event)
+
+            # TODO: Make all these Log Events into State so that we can get rid of the Event all together
+
+            action: Action = self.agent.step(self.state)
+            observation: str = action.run(self)
```

Comment:
Yes! In `State`. In the current implementation of the control loop, if I am reading this correctly, even if we stream multiple events to the Agent, the agent would still be blocked at the previous job before they can handle the next event. 
Instead of dispatching events to agents in real-time (like the `add_event`), we could potentially just maintain a list of Event/Observation history and pass it all together to the Agent when they finished with their previous action?

Reply:
Makes sense!

How will the agent collate Actions and Observations? E.g. how does it know the order of:
* I did this `runcommand` action
* Then I saw output X
* Then I read file foo.txt
* Then I saw output Y
* Then user told me to change tactics
* Then I thought Z

Does state keep a single array of those things?

Reply:
Yes -- that's my plan, `State` just keep a single array of those histories, and pass it to the agent every turn (i.e., it is just the `messages` argument for OpenAI API - you have to pass in the complete history to the model for it to emit the next response).

Reply:
OK great! So we have an array like the above--I assume they're all the same class? Are they all Actions?

Reply:
YES! Or just a list of Action and Observation (- both could be converted to string, or we could make Observation a separate class, but I'm still debating atm)

Reply:
That make a lot of sense! I'll figure it out :) Will let know when i finished this one!